self_test/generic/complex_matrices_test/cr_complex_matrix_test.cc

```diff
@@ -0,0 +1,228 @@
+  print_complex_matrix(matrix_square);
+
+  // test LU decomposition
+  cout << matrix_square.ludecompose() << endl;
+
+  constexpr unsigned long vector_length = n_row_square;
+
+  Vector<complex<double>> rhs(vector_length);
+  rhs[0] = complex<double>(1.0, 0.0);
+  rhs[1] = complex<double>(0.0, 1.0);
+  rhs[2] = complex<double>(2.0, 1.0);
+  rhs[3] = complex<double>(-3.0, 1.0);
+
+  // test residual
+  matrix_square.lubksub(rhs);
+  print_complex_vector(rhs);
+
+  Vector<complex<double>> x(vector_length);
+  x[0] = complex<double>(2.0, 2.0);
+  x[1] = complex<double>(-2.0, 3.0);
+  x[2] = complex<double>(0.5, 2.2);
+  x[3] = complex<double>(-1.4, 3.0);
+
+  Vector<complex<double>> soln(vector_length);
+
+  // test multiply
+  matrix_square.multiply(x, soln);
+  print_complex_vector(soln);
+
+  // test multiply transposed
+  matrix_square.multiply_transpose(x, soln);
+  print_complex_vector(soln);
+
+  LinearAlgebraDistribution* dist_pt = new LinearAlgebraDistribution();
+
+  constexpr unsigned n_nz_double = 5;
+
+  Vector<double> value_double(n_nz_double);
+  value_double[0] = 1.0;
+  value_double[1] = 2.0;
+  value_double[2] = 3.0;
+  value_double[3] = 4.0;
+  value_double[4] = 5.0;
+
+  Vector<int> column_index_double(n_nz_double);
+  column_index_double[0] = 0;
+  column_index_double[1] = 2;
+  column_index_double[2] = 0;
+  column_index_double[3] = 1;
+  column_index_double[4] = 1;
+
+  Vector<int> row_start_double(n_col_square + 1);
+  row_start_double[0] = 0;
+  row_start_double[1] = 2;
+  row_start_double[2] = 3;
+  row_start_double[3] = 4;
+  row_start_double[4] = n_nz_double;
+
+  CRDoubleMatrix matrix_in_double(
+    dist_pt, n_col_square, value_double, column_index_double, row_start_double);
+
+  // test add (CRDoubleMatrix)
+  matrix_square.add(matrix_in_double, matrix_square);
+  print_complex_matrix(matrix_square);
+
+  // test add (CRComplexMatrix)
+  matrix_square.add(matrix_square, matrix_square);
+  print_complex_matrix(matrix_square);
+
+  // Print input matrices
+  cout << "A" << endl;
+  print_complex_matrix(matrix_square);
+
+  CRComplexMatrix matrix_square_2(values_square,
+                                  column_index_square,
+                                  row_start_square,
+                                  n_col_square,
+                                  n_row_square);
+
+  cout << "B" << endl;
+  print_complex_matrix(matrix_square_2);
+
+  CRComplexMatrix matrix_result(values_square,
+                                column_index_square,
+                                row_start_square,
+                                n_col_square,
+                                n_row_square);
+
+  // test multiply method 2 (default)
+  matrix_square.multiply(matrix_square_2, matrix_result);
+  cout << "A B" << endl;
+  print_complex_matrix(matrix_result);
+
+  // Create a reference to Serial_matrix_matrix_multiply_method
+  unsigned& method = matrix_square.serial_matrix_matrix_multiply_method();
```

Comment:
Why do you use a reference to an unsigned here?

Reply:
See below